### PR TITLE
Preparation for Threshold plugin integration with Functions

### PR DESCRIPTION
**Pull request by KuphJr**
This PR contains all the changes required for the Functions Threshold plugin integration that don't depend on the custom config poller (See #9648).

In addition to the code to support the Threshold plugin integration, it includes some small bug fixes for the external adapter client, as well as a correction to `secrets.md` and `secrets.toml` for the name of the threshold key var.

### CODE REVIEW

File: core/services/ocr2/plugins/threshold/plugin.go
Comment (KuphJr) [inline core/services/ocr2/plugins/threshold/plugin.go:33]: We need a better approach here.  Do we think it is worth modifying the plugin directly to eliminate this map entirely in favor of using the indices from the `transmitters` array?
That would require future users of the plugin to also assign key shares in the same ordering as the `transmitters` array, however this seems to be the most straightforward approach.

File: core/services/functions/request.go
Comment (KuphJr) [inline core/services/functions/request.go:3]: The CBOR payload from on-chain has secrets encoded as bytes, but the adapter expects the secrets to be base64 string.  This is resolved by adding the new AdapterRequestData struct

File: core/services/functions/listener.go
Comment (bolekk) [inline core/services/functions/listener.go:374]: If it's not an error now then change to Warn or Info. Also remove the commented-out code.

File: core/services/functions/listener.go
Comment (bolekk) [inline core/services/functions/listener.go:366]: There should be an enum in gethwrappers instead of "1"

File: core/services/ocr2/plugins/functions/plugin.go
Comment (bolekk) [inline core/services/ocr2/plugins/functions/plugin.go:74]: Planning to do as part of this PR? If not, leave a JIRA ticket ID.

File: core/services/ocr2/plugins/threshold/decryption_queue.go
Comment (bolekk) [inline core/services/ocr2/plugins/threshold/decryption_queue.go:65]: remove

File: core/services/ocr2/plugins/threshold/plugin.go
Comment (bolekk) [reply core/services/ocr2/plugins/threshold/plugin.go:33]: yes

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/services/functions/request.go
Comment (bolekk) [reply core/services/functions/request.go:3]: Wait, that means the current RequestData is incorrect, right? I'll change it to bytes and add a test for that field, my bad!

File: core/services/functions/listener.go
Comment (bolekk) [reply core/services/functions/listener.go:366]: Looks like there isn't...

Comment (KuphJr) [reply core/services/functions/listener.go:366]: I see you added this const value in your PR.  Will rebase & use that value once your PR is merged.

File: core/services/functions/listener.go
Comment (KuphJr) [reply core/services/functions/listener.go:374]: This is an error.  If the secrets are simply invalid, then the response will be a userError, however an err indicates something is wrong w/ the adapter

File: core/services/functions/request.go
Comment (KuphJr) [reply core/services/functions/request.go:3]: As soon as that PR is merged, I'll rebase and remove the additional struct

File: core/services/functions/external_adapter_client.go
Comment (KuphJr) [inline core/services/functions/external_adapter_client.go:146]: This will be eliminated once I rebase changes from #9677

File: core/services/ocr2/plugins/s4/messages.proto
Comment (KuphJr) [inline core/services/ocr2/plugins/s4/messages.proto:1]: This caused a name conflict w/ the decryption plugin

File: core/services/functions/listener_test.go
Comment (bolekk) [inline core/services/functions/listener_test.go:63]: You don't need those huge real values here since you're mocking everything. Just create fake ones that are much smaller.

File: core/services/ocr2/delegate.go
Comment (bolekk) [inline core/services/ocr2/delegate.go:416]: mark the "1" as pluginID, either with an extra var or a comment

File: core/services/ocr2/plugins/functions/config/config.go
Comment (bolekk) [inline core/services/ocr2/plugins/functions/config/config.go:26]: Put in a sub-struct please. Otherwise it's too confusing.
I see you already have a struct in plugin.go, just put it here.

File: core/services/ocr2/plugins/functions/plugin.go
Comment (bolekk) [inline core/services/ocr2/plugins/functions/plugin.go:50]: Andrei renamed this field to QConfig, don't change it back.

File: core/services/ocr2/plugins/threshold/plugin.go
Comment (bolekk) [inline core/services/ocr2/plugins/threshold/plugin.go:57]: I still don't get why you can't just put those types here. Unmarshal literally does what you're doing below for all fields. Can you explain that again, please?

File: integration-tests/go.sum
Comment (bolekk) [inline integration-tests/go.sum:1387]: Is this needed?

File: core/services/functions/listener.go
Comment (bolekk) [inline core/services/functions/listener.go:381]: Maybe avoid "error" in the message. How about "no threshold decryption detected, falling back to legacy secrets" or something along those lines?

File: core/services/functions/listener.go
Comment (bolekk) [inline core/services/functions/listener.go:388]: hmmm is this already base64? Or is a json..? I'm confused because we put it directly into a string field..

File: core/services/functions/listener.go
Comment (bolekk) [inline core/services/functions/listener.go:383]: Is this expected to be always false until be fully integrate and provide valid config?

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/scripts/go.mod
Comment (KuphJr) [reply core/scripts/go.mod:297]: I think this is just bumping tdh2 to the most recent commit, which is probably what we want, right?

File: core/services/functions/listener_test.go
Comment (KuphJr) [reply core/services/functions/listener_test.go:162]: Not on purpose.  Might have been a merge conflict issue.

File: core/services/functions/listener_test.go
Comment (KuphJr) [reply core/services/functions/listener_test.go:174]: Correct.  They are not actually needed to run the test here.

File: core/services/ocr2/plugins/threshold/plugin.go
Comment (KuphJr) [reply core/services/ocr2/plugins/threshold/plugin.go:57]: There is additional logic required to unmarshal the JSON into the tdh2easy.PublicKey & tdh2easy.PrivateKey types beyond a simple JSON unmarshal.

File: integration-tests/go.sum
Comment (KuphJr) [reply integration-tests/go.sum:1387]: This is updating TDH2 to the latest commit.

File: core/services/functions/listener.go
Comment (KuphJr) [reply core/services/functions/listener.go:383]: Correct.  The threshold encrypted secrets payload which is stored by the user off-chain has a particular format which can be distinguished from DON-key encrypted secrets.  If the secrets are not in threshold format, then the adapter returns a user error during the fetch attempt.

File: core/scripts/go.mod
Comment (bolekk) [reply core/scripts/go.mod:297]: What I meant is that you already have tdh2 in top-level go.mod and go.sum. You don't need it in core/scripts/go.mod or integration-tests/go.mod at all, correct?

File: core/services/functions/listener.go
Comment (KuphJr) [reply core/services/functions/listener.go:369]: No, this is correct.  nodeProvidedSecrets is a JSON string, not a base64 encoded bytes.  The structure of nodeProvidedSecrets is as follows

```
{
  "0x": "base64string",
  "0xNodeAddress0": "base64string w/ per-node assigned secret",
  ...
}
```

Comment (bolekk) [reply core/services/functions/listener.go:369]: That's not great. We have a JSON saved inside a string field of a top-level JSON... Maybe we should base64-encode the whole thing again?

File: core/scripts/go.mod
Comment (KuphJr) [reply core/scripts/go.mod:297]: You do.  This was automatic when I ran `make gomodtidy`

File: core/services/functions/listener.go
Comment (KuphJr) [reply core/services/functions/listener.go:369]: That would require a change to the EA.  LMK if you think this should be completed before this PR is merged.
Else, I can include this change in the final Threshold integration PR so I can work on the adapter PR while this PR makes its way thru the merge queue.

Comment (bolekk) [reply core/services/functions/listener.go:369]: Sure we can change separately.

File: core/services/ocr2/plugins/threshold/plugin.go
Comment (bolekk) [reply core/services/ocr2/plugins/threshold/plugin.go:57]: Yes but that logic sits in PublicKey.Unmarshal() and that should be called automatically for all fields if you unmarshal the struct. Can you confirm that it works (or doesn't)?

File: core/services/ocr2/plugins/functions/config/config.go
Comment (KuphJr) [reply core/services/ocr2/plugins/functions/config/config.go:26]: My goal was to avoid having to nest these config values under another field in the jobspec TOML.  However, to make the code more readable I used embedding.

File: core/services/ocr2/plugins/threshold/plugin.go
Comment (KuphJr) [reply core/services/ocr2/plugins/threshold/plugin.go:57]: After a quick search, it appears this is mostly correct.  json.Unmarshal will call the UnmarshalJSON method for the types contained within the struct.

Unfortunately, it appears tdh2easy uses the method name `Unmarshal` instead of `UnmarshalJSON`, so that will not work in this case. We need to get a PR merged to tdh2easy to change the method name `Unmarshal` to `UnmarshalJSON`.

Is is possible that we could do this in the final integration PR to avoid blocking this PR on the necessary change to tdh2?

Comment (bolekk) [reply core/services/ocr2/plugins/threshold/plugin.go:57]: sure, np

File: core/services/functions/listener_test.go
Comment (bolekk) [inline core/services/functions/listener_test.go:192]: Here you can validate that DecryptedSecrets were passed to RunComputation

File: core/services/ocr2/plugins/functions/config/config.go
Comment (bolekk) [inline core/services/ocr2/plugins/functions/config/config.go:26]: Better to go with nested fields, like for the gateway connector config. S4 will be another nested one.

File: core/services/functions/listener.go
Comment (bolekk) [inline core/services/functions/listener.go:370]: In this PR, decryptor is always nil, correct?

File: core/services/functions/listener.go
Comment (bolekk) [reply core/services/functions/listener.go:383]: Agreed to address separately.

File: core/services/functions/listener_test.go
Comment (bolekk) [inline core/services/functions/listener_test.go:102]: How come this doesn't break the test named "HandleOracleRequestCBORParsingCorrect"? That test has non-empty secrets so it should make calls FetchEncryptedSecrets and Decrypt...

I'm confused.

File: core/services/ocr2/plugins/functions/plugin.go
Comment (bolekk) [inline core/services/ocr2/plugins/functions/plugin.go:166]: Those validations should go into ValidatePluginConfig() in config.go (which currently does nothing..) When you move Threshold config to a nested struct that can be nil, you can validate values only if they exist and skip otherwise.

You can also add the defaults over there as consts or maybe better require all values to be set explicitly?

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/services/functions/listener.go
Comment (KuphJr) [reply core/services/functions/listener.go:370]: Correct.

File: core/services/ocr2/plugins/functions/plugin.go
Comment (KuphJr) [reply core/services/ocr2/plugins/functions/plugin.go:166]: I prefer default values at least for now, then maybe we can deprecate if we decide we don't want defaults.

File: core/services/functions/listener_test.go
Comment (KuphJr) [reply core/services/functions/listener_test.go:102]: This test is missing the SecretsLocation param and therefor fails the check `requestData.SecretsLocation == LocationRemote`

**Review by bolekk — APPROVED** [review]

**Pull request merged**
